Add files via upload

CodeDNA opened the pull request:
I am uploading a code and the explanation of the code(word file).
Please have a look and work on it for improvements.

CodeDNA committed:
Add files via upload

CodeDNA committed:
Create PREDICTION_FOR_REPETITION

devoshm commented on PREDICTION_FOR_REPETITION:
```diff
@@ -0,0 +1,66 @@
+		System.out.println("Total repetitions :"+count);
+		System.out.println("Percentage of repetitions: "+Percentage+"%");
+		
+		if(Percentage>Threshold*100) 
```
We shouldn't be getting the threshold from the users. We must be able to set a threshold based on array pattern.

CodeDNA replied:
Yes! we are not going to get the 'threshold' value from the end user. Obviously, the user doesn't have to know about anything but the array he is trying to sort.  Here the threshold (and the size of the array) is used by us to check various test cases and decide the threshold range where the sorting algorithms, that are good for repetition arrays, work best. [I would request you to please read the word file also, where the working is explained.] :)

devoshm commented on PREDICTION_FOR_REPETITION:
```diff
@@ -0,0 +1,66 @@
+		arraysize= InputSize.nextInt();
+		
+		DataGenerator dataGenerator = new DataGenerator(DataGenerator.PATTERN_REPEATED, arraysize);
+    System.out.println(dataGenerator.generate());
```
Calling dataGenerator.generate() would create random array each time it is called. Therefore, the list you are printing here is not the same as the one that is generated in the next line. Hope you are aware of it.

CodeDNA replied:
Yes! You are right. I missed that. 👍 
So, earlier the array was printed using the - "dataGenerator.generate()" function so it will always generate and print a different array as the previous line.
This will be solved by using "GetArray" list for printing, which is used to store the generated array.
Bingo! :)

CodeDNA committed:
Update PREDICTION_FOR_REPETITION

CodeDNA committed:
Update PREDICTION_FOR_REPETITION

CodeDNA committed:
Update PREDICTION_FOR_REPETITION

ARRAY TEST FOR REPEATED ELEMENTS.

How it works?
We define a threshold for percentage of repetitions in the array. If the percentage of repetition is above the threshold percentage we define it having the pattern as ‘REPEATED’, which means that it has many elements that are repeated and the sorting algorithm that are good for array with repeated elements can be applied. Examples of algorithms that are good for this type of array includes: Quick sort, Shell Sort, Heap Sort. 

IMPLEMENTATION:
Array generation:
1)	Get the array using ‘DataGenerator’ class. Here I have saved in the test array-‘TArr’.
2)	‘TestReps’ is the function that takes the array and its size as input.


Functioning of ‘TestReps()’:
Declarations:
1)	‘threshold’ is used for storing the threshold percentage.
2)	‘count’ acts as the counter for the number of repetitions encountered.
3)	‘percentage’ is used to store the percentage of repetitions encountered.
4)	‘length’  is the array size.
Working:
We create a HASHSET to store the numbers of the array. Hash sets have the property that it doesn’t contain duplicate values. When we try to add a value that is already in the hash set, it is overwritten and the add() function returns false since nothing was appended.
Whenever the add() function returns ‘false’ we increment the  ‘count’ by 1 and add() function will return false every time it encounters a duplicate value from the array. Finally, the ‘count’ will give the number of repetitions in the array.
Percentage of repetitions = (Total repetitions/Total Numbers in the array)*100
[‘length’ is the length of array]

Now we compare the percentage obtained by the above formula with the threshold percentage.
Percentage > Threshold :  Pattern of the array is “REPEATED”.
Percentage < Threshold :  Pattern of the array is “NOT REPEATED”

CodeDNA committed:
Description for the "PREDICTION_FOR_REPETITION"

CodeDNA committed:
Delete PREDICTION_FOR_REPETITION

CodeDNA committed:
Create PREDICTION_FOR_REPETITION

CodeDNA committed:
Delete ARRAY TEST FOR REPEATED ELEMENTS.docx
